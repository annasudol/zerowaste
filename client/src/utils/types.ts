export interface LocationTypes {
    hash?: string;
    pathname: string;
    search?: string;
    state: {
        ingredients: string[];
    };
    key?: string;
}
