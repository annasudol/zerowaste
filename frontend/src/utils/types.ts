export interface LocationTypes {
    hash?: string
    pathname: string
    search?: string
    state: {
        ingredients?: string[]
        backPath?: string
        callRefetch?: boolean
    };
    key?: string
}



export interface ListType {
    title: string
    id: number
}


export interface RecipeStateProps {
    title: string
    servings: number
    image: string
    readyInMinutes: number
    ingredients: string[] | []
    detailedIngredients: string[] | []
    instructions: string
    sourceUrl: undefined | string
}