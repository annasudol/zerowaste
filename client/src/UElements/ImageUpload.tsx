import * as React from "react";

import { Button } from '../UElements';
// import './ImageUpload.css';

interface ImageUploadProps {
    onInput?(id: string, pickedFile: string, fileIsValid: boolean): VoidFunction
    id?: string
}
export const ImageUpload: React.FunctionComponent<ImageUploadProps> = ({ onInput, id }): React.ReactElement => {
    const [file, setFile] = React.useState();
    const [previewUrl, setPreviewUrl] = React.useState();
    const [isValid, setIsValid] = React.useState(false);

    const filePickerRef: any = React.useRef();

    React.useEffect(() => {
        if (file) {
            return;
        }
        const fileReader: any = new FileReader();
        fileReader.onload = () => {
            setPreviewUrl(fileReader.result);
        };
        fileReader.readAsDataURL(file);
    }, [file]);

    const pickedHandler = event => {
        let pickedFile;
        let fileIsValid = isValid;
        if (event.target.files && event.target.files.length === 1) {
            pickedFile = event.target.files[0];
            setFile(pickedFile);
            setIsValid(true);
            fileIsValid = true;
        } else {
            setIsValid(false);
            fileIsValid = false;
        }
        // onInput(id, pickedFile, fileIsValid);
    };

    const pickImageHandler = () => {
        filePickerRef.current.click();
    };

    return (
        <div className="form-control">
            <input
                id={id}
                ref={filePickerRef}
                style={{ display: 'none' }}
                type="file"
                accept=".jpg,.png,.jpeg"
                onChange={pickedHandler}
            />
            <div className="image-upload center">
                <div className="image-upload__preview">
                    {previewUrl && <img src={previewUrl} alt="Preview" />}
                    {!previewUrl && <p>Please pick an image.</p>}
                </div>
                <Button onClick={pickImageHandler}>
                    PICK IMAGE
                </Button>
            </div>
            {!isValid && <p>Error with uploading image</p>}
        </div>
    );
};