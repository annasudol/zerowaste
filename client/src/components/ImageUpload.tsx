import * as React from 'react';
import { Upload, message } from 'antd';
import { PlusOutlined } from '@ant-design/icons';

export interface FileType {
    lastModified: number
    name: string
    size: number
    type: "image/jpeg"
    uid: string
}
interface ImageUploadProps {
    imageUrl: string | undefined
    form?(file: FileType): void
}

function getBase64(img: any, callback: any) {
    const reader = new FileReader();
    reader.addEventListener('load', () => callback(reader.result));
    reader.readAsDataURL(img);
}

function beforeUpload(file) {
    const isJpgOrPng = file.type === 'image/jpeg' || file.type === 'image/png';
    if (!isJpgOrPng) {
        message.error('You can only upload JPG/PNG file!');
    }
    const isLt2M = file.size / 1024 / 1024 < 2;
    if (!isLt2M) {
        message.error('Image must smaller than 2MB!');
    }
    return isJpgOrPng && isLt2M;
}

export const ImageUpload: React.FC<ImageUploadProps> = ({ imageUrl, form }): React.ReactElement => {
    const [img, setImg] = React.useState<string | undefined>()
    const [file, setFile] = React.useState<any>()


    const handleChange = (info) => {
        if (info.file.originFileObj) {
            // tslint:disable-next-line: no-shadowed-variable
            getBase64(info.file.originFileObj, (imageUrl: string) => {
                setImg(imageUrl);
                if (form) {
                    setFile(info.file.originFileObj);
                    form(info.file.originFileObj);
                }
            })
        }
    }



    const uploadButton = (
        <div>
            <PlusOutlined />
            <div className='ant-upload-text'>Upload</div>
        </div>
    );
    return (
        <>
            <Upload
                name="avatar"
                listType="picture-card"
                className="avatar-uploader"
                showUploadList={false}
                action={file}
                beforeUpload={beforeUpload}
                onChange={handleChange}

            >
                {img || imageUrl ? <img src={img || imageUrl} alt="avatar" style={{ width: '100%' }} /> : uploadButton}
            </Upload>

        </>
    )
};