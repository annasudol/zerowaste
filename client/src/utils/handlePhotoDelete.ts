export const handlePhotoDelete = async (url: string) => {
    const publicId = url.match(/(?=).{20}?(?=.jpe?g|.png)/);
    if (publicId) {
        try {
            const response = await fetch('http://localhost:9000/delete/', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({ public_id: publicId[0] })
            });
            return response.json();

        } catch { }
    }

}