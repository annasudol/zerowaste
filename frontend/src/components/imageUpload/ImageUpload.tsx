import React, { FC, ReactElement } from 'react';
import { Upload, message } from 'antd';
import { PlusOutlined } from '@ant-design/icons';
import { UploadChangeParam } from "antd/lib/upload";
import { UploadFile } from "antd/lib/upload/interface";


interface ImageUploadProps {
    imageUrl?: string
    form?(file: Blob): void
}

function getBase64(img: Blob, callback: any) {
    const reader = new FileReader();
    reader.addEventListener('load', () => callback(reader.result));
    reader.readAsDataURL(img);
}

function beforeUpload(file: { type: string; size: number; }) {
    const isJpgOrPng = file.type === 'image/jpeg' || file.type === 'image/png';
    if (!isJpgOrPng) {
        message.error('You can only upload JPG/PNG file!');
    }
    const isLt2M = file.size / 1024 / 1024 < 2;
    if (!isLt2M) {
        message.error('Image must smaller than 2MB!');
    }
    return isJpgOrPng && isLt2M;
}

export const ImageUpload: FC<ImageUploadProps> = ({ imageUrl, form }): ReactElement => {
    const [img, setImg] = React.useState<string | undefined>()
    const [file, setFile] = React.useState<any>()

    const handleChange = (info: UploadChangeParam<UploadFile<any>>) => {
        const file =info?.file?.originFileObj

         if (file) {
            getBase64(file, (imageUrl: string) => {
                setImg(imageUrl);
                if (form) {
                    setFile(file);
                    form(file);
                }
            })
        }
    }




    const uploadButton = (
        <div>
            <PlusOutlined />
            <div className='ant-upload-text'>Upload</div>
        </div>
    );
    return (
        <>
            <Upload
                name="avatar"
                listType="picture-card"
                className="avatar-uploader"
                showUploadList={false}
                action={file}
                beforeUpload={beforeUpload}
                onChange={(info)=> handleChange(info)}

            >
                {img || imageUrl ? <img src={img || imageUrl} alt="avatar" style={{ width: '100%' }} /> : uploadButton}
            </Upload>

        </>
    )
};