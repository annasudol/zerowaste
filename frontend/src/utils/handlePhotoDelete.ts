export const handlePhotoDelete = async (url: string) => {
    const publicId = url.match(/(?=).{20}?(?=.jpe?g|.png)/);
    if (publicId) {
        try {
            const response = await fetch(`http://localhost:${process.env.REACT_APP_PORT}/delete/`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({ public_id: publicId[0] })
            });
            console.log(response, "response")
            return response.json();

        } catch { }
    }

}