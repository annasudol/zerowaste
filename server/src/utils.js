
module.exports.paginateResults = ({
  page = 0,
  pageSize = 10,
  results,
}) => {
  const resultsIntoPages = page * pageSize;
  return results.slice(resultsIntoPages, resultsIntoPages + pageSize);
};
