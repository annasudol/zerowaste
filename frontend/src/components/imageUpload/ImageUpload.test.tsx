import { shallow } from 'enzyme';
import React from 'react';
import { Upload } from 'antd';
import { ImageUpload } from './ImageUpload';
import { renderReact } from "../../test.utils";
import {screen} from '@testing-library/react'
import toJson from 'enzyme-to-json';


describe('component', (): void => {
  describe('ImageUpload', (): void => {
    it('match snapshot', (): void => {
      const wrapper = shallow(<ImageUpload />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('renders image with correct src', (): void => {
        const imageUrl = "https://res.cloudinary.com/drgb4slzt/image/upload/v1593443213/belhdjaq4pevmwydoept.jpg"
        renderReact(<ImageUpload imageUrl={imageUrl} form={jest.fn()} />);
        expect(screen.getByAltText('avatar')).toHaveAttribute("src", imageUrl);
      });

    it('match snapshot when given correct file format', ()=> {
      const file = {
        lastModified: 1592474852973,
        lastModifiedDate: 'Thu Jun 18 2020 12:07:32 GMT+0200 (Central European Summer Time)',
        name: "16800-blue-ribbon-apple-pie-600x600.jpg",
        originFileObj: new Blob([JSON.stringify({
          lastModified: 1592474852973,
          name: "16800-blue-ribbon-apple-pie-600x600.jpg",
          size: 128209,
          type: "image/jpeg",
          uid: "rc-upload-1593444261913-2",
          webkitRelativePath: "",
        })], {type : 'text/html'}),
        percent: 100,
        size: 128209,
        status: "uploading",
        type: "image/jpeg",
        uid: "rc-upload-1593444261913-2",
      };
      const beforeUpload = {
        lastModified: 1592474852973,
        name: "16800-blue-ribbon-apple-pie-600x600.jpg",
        size: 128209,
        type: "image/jpeg",
        uid: "rc-upload-1593597112355-2",
        webkitRelativePath: ""
      }
      const wrapper = shallow(<ImageUpload form={jest.fn()} />)
      wrapper.find(Upload).simulate('beforeUpload', {file});
      wrapper.find(Upload).prop<Function>('beforeUpload')(beforeUpload);

      wrapper.find(Upload).simulate('change', {file});
      wrapper.update();
      expect(toJson(wrapper)).toMatchSnapshot();

    });

    it('match snapshot when wrong file format was given', ()=> {
      const beforeUpload = {
        lastModified: 1590938637311,
        name: "now.json",
        size: 143,
        type: "application/json",
        uid: "rc-upload-1593597112355-4",
        webkitRelativePath: "",
      }

      const wrapper = shallow(<ImageUpload form={jest.fn()} />)
      wrapper.find(Upload).prop<Function>('beforeUpload')(beforeUpload);
      wrapper.update();

      expect(toJson(wrapper)).toMatchSnapshot();


    })

    it('match snapshot when too large image was given', ()=> {
      const beforeUpload = {
        lastModified: 1592474852973,
        name: "16800-blue-ribbon-apple-pie-600x600.jpg",
        size: 12820900,
        type: "image/jpeg",
        uid: "rc-upload-1593597112355-2",
        webkitRelativePath: ""
      }

      const wrapper = shallow(<ImageUpload form={jest.fn()} />)
      wrapper.find(Upload).prop<Function>('beforeUpload')(beforeUpload);
      wrapper.update();

      expect(toJson(wrapper)).toMatchSnapshot();
    })
  });
});