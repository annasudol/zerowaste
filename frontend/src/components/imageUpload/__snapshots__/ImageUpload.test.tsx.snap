// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component ImageUpload match snapshot 1`] = `
<Fragment>
  <Upload
    accept=""
    action=""
    beforeUpload={[Function]}
    className="avatar-uploader"
    data={Object {}}
    disabled={false}
    listType="picture-card"
    multiple={false}
    name="avatar"
    onChange={[Function]}
    showUploadList={false}
    supportServerRender={true}
    type="select"
  >
    <ForwardRef(PlusOutlined) />
    <div
      className="ant-upload-text"
    >
      Upload
    </div>
  </Upload>
</Fragment>
`;

exports[`component ImageUpload match snapshot when given correct file format 1`] = `
<Fragment>
  <Upload
    accept=""
    action=""
    beforeUpload={[Function]}
    className="avatar-uploader"
    data={Object {}}
    disabled={false}
    listType="picture-card"
    multiple={false}
    name="avatar"
    onChange={[Function]}
    showUploadList={false}
    supportServerRender={true}
    type="select"
  >
    <ForwardRef(PlusOutlined) />
    <div
      className="ant-upload-text"
    >
      Upload
    </div>
  </Upload>
</Fragment>
`;

exports[`component ImageUpload match snapshot when too large image was given 1`] = `
<Fragment>
  <Upload
    accept=""
    action=""
    beforeUpload={[Function]}
    className="avatar-uploader"
    data={Object {}}
    disabled={false}
    listType="picture-card"
    multiple={false}
    name="avatar"
    onChange={[Function]}
    showUploadList={false}
    supportServerRender={true}
    type="select"
  >
    <ForwardRef(PlusOutlined) />
    <div
      className="ant-upload-text"
    >
      Upload
    </div>
  </Upload>
</Fragment>
`;

exports[`component ImageUpload match snapshot when wrong file format was given 1`] = `
<Fragment>
  <Upload
    accept=""
    action=""
    beforeUpload={[Function]}
    className="avatar-uploader"
    data={Object {}}
    disabled={false}
    listType="picture-card"
    multiple={false}
    name="avatar"
    onChange={[Function]}
    showUploadList={false}
    supportServerRender={true}
    type="select"
  >
    <ForwardRef(PlusOutlined) />
    <div
      className="ant-upload-text"
    >
      Upload
    </div>
  </Upload>
</Fragment>
`;
